pkg/reconciler/workload/resource/resource_controller.go

```diff
@@ -70,10 +73,56 @@ func NewController(
 
 		dynClusterClient: dynamicClusterClient,
 
-		namespaceLister: namespaceInformer.Lister(),
+		getNamespace: func(clusterName logicalcluster.Name, namespaceName string) (*corev1.Namespace, error) {
+			namespace, err := namespaceInformer.Lister().Get(kcpcache.ToClusterAwareKey(clusterName.String(), "", namespaceName))
+			if err != nil {
+				return nil, err
+			}
+			// Listers are not workspaced, so we need to filter out namespaces that don't match the workspace.
+			if logicalcluster.From(namespace) != clusterName {
+				return nil, nil
+			}
+			return namespace, err
+		},
 
-		syncTargetLister:  syncTargetInformer.Lister(),
-		syncTargetIndexer: syncTargetInformer.Informer().GetIndexer(),
+		getValidSyncTargetKeysForWorkspace: func(clusterName logicalcluster.Name) (sets.String, error) {
+			placements, err := placementInformer.Informer().GetIndexer().ByIndex(byLocationWorkspace, clusterName.String())
+			if err != nil {
+				return nil, err
+			}
+
+			expectedSyncTargetKeys := sets.String{}
+			for _, placement := range placements {
+				placement := placement.(*schedulingv1alpha1.Placement)
+				if val, ok := placement.Annotations[workloadv1alpha1.InternalSyncTargetPlacementAnnotationKey]; ok && val != "" {
+					labelSelector := labels.SelectorFromSet(map[string]string{workloadv1alpha1.InternalSyncTargetKeyLabel: val})
+					syncTargets, err := syncTargetInformer.Lister().List(labelSelector)
+					if err != nil {
+						return nil, err
+					}
+					for _, syncTarget := range syncTargets {
+						expectedSyncTargetKeys.Insert(workloadv1alpha1.ToSyncTargetKey(logicalcluster.From(syncTarget), syncTarget.Name))
+					}
+				}
+			}
+			return expectedSyncTargetKeys, err
+		},
+
+		getSyncTargetFromKey: func(syncTargetKey string) (*workloadv1alpha1.SyncTarget, error) {
```

Comment:
I'd recommend returning *SyncTarget, found bool, error. Getting back `nil, nil` is maybe a bit "weird" 😄

Reply:
done